### PR TITLE
Add bound check to int() and update tests for type casting

### PR DESCRIPTION
**Pull request by Aiee**
As title.

### CODE REVIEW

**Review by jievince — DISMISSED** [review]

File: src/common/datatypes/Value.cpp [hunk 1]
```diff
@@ -1610,6 +1610,11 @@ std::pair<int64_t, bool> Value::toInt() {
             return std::make_pair(getInt(), true);
         }
         case Value::Type::FLOAT: {
+            // Check if float value is in the limit of int_64
+            if (getFloat() < std::numeric_limits<int64_t>::min() ||
+                getFloat() > std::numeric_limits<int64_t>::max()) {
+                    return std::make_pair(int64_t{}, false);
```
Comment (Shylock-Hg) [inline src/common/datatypes/Value.cpp:1616]: The cypher will return min/max value of int64_t when overflow.

Comment (czpmango) [inline src/common/datatypes/Value.cpp:1616]: Why add this logic? The numerical value will be truncated when converted to int.

Comment (Aiee) [reply src/common/datatypes/Value.cpp]: Thanks for pointing that. I have updated the value returned in a overflow.

Comment (Aiee) [reply src/common/datatypes/Value.cpp]: I believe the user should be somehow informed when an overflow happens. Returning a truncated value without notifying the user may cause confusion.

Comment (Shylock-Hg) [reply src/common/datatypes/Value.cpp]: static_cast is ok for this logical. You could remain the tests.

File: src/common/datatypes/Value.cpp [hunk 2]
```diff
@@ -1610,6 +1610,16 @@ std::pair<int64_t, bool> Value::toInt() {
             return std::make_pair(getInt(), true);
         }
         case Value::Type::FLOAT: {
+            // Check if float value is in the limit of int_64
+            if (getFloat() < std::numeric_limits<int64_t>::min()) {
+                LOG(WARNING) << "Overflow happened when casting float value " << getFloat()
```
Comment (Shylock-Hg) [inline src/common/datatypes/Value.cpp:1615]: And don't log in so low-level layer it maybe hurt performance.

File: src/common/datatypes/Value.cpp (hunk 1 above)
Comment (czpmango) [reply src/common/datatypes/Value.cpp]: I think your consideration is reasonable.
This is my test in mysql:
```
mysql> select CONVERT((-9223372036854775809.3),unsigned);
+--------------------------------------------+
| CONVERT((-9223372036854775809.3),unsigned) |
+--------------------------------------------+
|                        9223372036854775808 |
+--------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-------------------------------------------------------------+
| Level   | Code | Message                                                     |
+---------+------+-------------------------------------------------------------+
| Warning | 1292 | Truncated incorrect DECIMAL value: '-9223372036854775809.3' |
+---------+------+-------------------------------------------------------------+
1 row in set (0.00 sec)
```

File: src/common/datatypes/Value.cpp (hunk 2 above)
Comment (czpmango) [reply src/common/datatypes/Value.cpp:1615]: > And don't log in so low-level layer it maybe hurt performance.

Agree.

File: src/common/datatypes/Value.cpp (hunk 1 above)
Comment (Aiee) [reply src/common/datatypes/Value.cpp]: So theoretically converting a float into an int is always valid, is it correct?

File: src/common/datatypes/Value.cpp (hunk 2 above)
Comment (Aiee) [reply src/common/datatypes/Value.cpp:1615]: Makes sense. I'll log the warning when calling this function.

File: src/common/datatypes/Value.cpp (hunk 1 above)
Comment (Aiee) [reply src/common/datatypes/Value.cpp]: Good to know about this. Thanks.

**Comment by Aiee** [comment]
How about this approach:
In an overflow, toInt() returns the ```truncated result``` as the first element of the pair, and ```false``` as the second so that we can log warnings base on this boolean value.

WDYT? @Shylock-Hg @czpmango

File: src/common/datatypes/Value.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/common/datatypes/Value.cpp]: Could keep the check.It's maybe a U.B.

**Comment by Shylock-Hg** [comment]
> How about this approach:
> In an overflow, toInt() returns the `truncated result` as the first element of the pair, and `false` as the second so that we can log warnings base on this boolean value.
> 
> WDYT? @Shylock-Hg @czpmango

Make sense.

File: src/common/datatypes/Value.cpp [hunk 3]
```diff
@@ -1610,6 +1610,12 @@ std::pair<int64_t, bool> Value::toInt() {
             return std::make_pair(getInt(), true);
         }
         case Value::Type::FLOAT: {
+            // Check if float value is in the limit of int_64
+            if (getFloat() < std::numeric_limits<int64_t>::min() ||
```
Comment (Shylock-Hg) [inline src/common/datatypes/Value.cpp:1614]: Less and equal

File: src/common/datatypes/Value.cpp [hunk 4]
```diff
@@ -1610,6 +1610,12 @@ std::pair<int64_t, bool> Value::toInt() {
...
+                getFloat() > std::numeric_limits<int64_t>::max()) {
```
Comment (Shylock-Hg) [inline src/common/datatypes/Value.cpp:1615]: ditto.

File: src/common/datatypes/Value.cpp [hunk 5]
```diff
@@ -1610,6 +1610,12 @@ std::pair<int64_t, bool> Value::toInt() {
...
+                // return a truncated value when overflow
+                return std::make_pair(static_cast<int64_t>(getFloat()), false);
```
Comment (Shylock-Hg) [inline src/common/datatypes/Value.cpp:1617]: Don't cast it, it's maybe a U.B.

Comment (Aiee) [reply src/common/datatypes/Value.cpp]: Fixed

File: src/common/datatypes/Value.cpp [hunk 6]
```diff
@@ -1610,6 +1610,12 @@ std::pair<int64_t, bool> Value::toInt() {
             return std::make_pair(getInt(), true);
         }
         case Value::Type::FLOAT: {
+            // Check if float value is in the range of int_64
+            if (getFloat() <= std::numeric_limits<int64_t>::min()) {
+                return std::make_pair(std::numeric_limits<int64_t>::min(), false);
```
Comment (Shylock-Hg) [inline src/common/datatypes/Value.cpp:1615]: return false means the conversion is invalid, but Cypher allow this conversion even if overflow.

Comment (Aiee) [reply src/common/datatypes/Value.cpp:1615]: if we return true for overflow then how could we distinguish an overflow and a non-overflow situation?

Comment (Shylock-Hg) [reply src/common/datatypes/Value.cpp:1615]: Will cypher do it?

Comment (Aiee) [reply src/common/datatypes/Value.cpp:1615]: Screen Shot 2021-03-02 at 11 43 25 AM

cypher returns the int_min/max without logging

Comment (Shylock-Hg) [reply src/common/datatypes/Value.cpp:1615]: Following the cypher.

**Review by Shylock-Hg — APPROVED** [review]

**Review by czpmango — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**